dotCMS/src/main/java/com/dotmarketing/common/reindex/IndexResource.java

```diff
@@ -0,0 +1,381 @@
+        }
+        return Response.ok(new ResponseEntityView(builder.build())).build();
+    }
+    
+    @CloseDBIfOpened
+    @GET
+    @JSONP
+    @NoCache
+    @Path("/failed")
+    @Produces({MediaType.APPLICATION_JSON})
+    public Response downloadRemainingRecordsAsCsv(@Context final HttpServletRequest request,
+                    @Context final HttpServletResponse response) throws DotDataException {
+        validateUser(request, response);
+        List<Map<String, Object>> results = new ArrayList<>();
+        
+        
+        
+        APILocator.getReindexQueueAPI().getFailedReindexRecords().stream().forEach(row->{
+            Map<String, Object> failure = new HashMap<>();
+            
+            final Contentlet contentlet = Try.of(()->APILocator.getContentletAPI().findContentletByIdentifierAnyLanguage(row.getIdentToIndex()))
+                            .onFailure(e->Logger.warn(IndexResource.class, e.getMessage(), e))
+                            .getOrElse(Contentlet::new);
+            Map<String, Object> conMap = Try.of(() -> new ContentletToMapTransformer(contentlet).toMaps().get(0))
+                            .onFailure(e->Logger.warn(IndexResource.class, e.getMessage(), e))
+                            .getOrElse(HashMap::new);
+            
+            failure.put("title", contentlet.getTitle());
+            failure.put("inode", contentlet.getInode());
+            failure.put("identifier", row.getIdentToIndex());
+            failure.put("serverId", row.getServerId());
+            failure.put("failureReason", row.getLastResult());
+            failure.put("priority", row.getPriority());
+            failure.put("contentlet", conMap);
+            results.add(failure);
+        });
+
+        return Response.ok(results).build();
+    }
+
+    @CloseDBIfOpened
+    @DELETE
+    @JSONP
+    @NoCache
+    @Path("/failed")
+    @Produces({MediaType.APPLICATION_JSON})
+    public Response deleteFailedRecords(@Context final HttpServletRequest request, @Context final HttpServletResponse response)
+                    throws DotDataException {
+        final InitDataObject init = validateUser(request, response);
+        APILocator.getReindexQueueAPI().deleteFailedRecords();
+        return Response.ok(new ResponseEntityView(true)).build();
+    }
+
+    private void sendAdminMessage(String message, MessageSeverity severity, long millis) {
+    
+    
+            final SystemMessageBuilder systemMessageBuilder = new SystemMessageBuilder();
+    
+    
+            
+            SystemMessage systemMessage = systemMessageBuilder.setMessage(message)
+                 .setType(MessageType.SIMPLE_MESSAGE)
+                 .setSeverity(severity)
+                 .setLife(millis)
+                 .create();
+             List<String> users = Try.of(()->APILocator.getRoleAPI().findUserIdsForRole(APILocator.getRoleAPI().loadCMSAdminRole())).getOrElse(ImmutableList.of());
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Local variable 'users' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5632525)